chainer/distributions/bernoulli.py

```diff
@@ -0,0 +1,119 @@
+import chainer
+from chainer.backends import cuda
+from chainer import distribution
+from chainer.functions.activation import sigmoid
+from chainer.functions.array import broadcast
+from chainer.functions.math import exponential
+import numpy
+
+
+class Bernoulli(distribution.Distribution):
+
+    """Bernoulli Distribution.
+
+    The probability mass function of the distribution is expressed as
+
+    .. math::
+        P(x = 1; p) = p \\\\
+        P(x = 0; p) = 1 - p
+
+    Args:
+        p(:class:`~chainer.Variable` or :class:`numpy.ndarray` or \
+        :class:`cupy.ndarray`): Parameter of distribution representing \
+        :math:`p`. Either `p` or `logit` (not both) must have a value.
+        logit(:class:`~chainer.Variable` or :class:`numpy.ndarray` or \
+        :class:`cupy.ndarray`): Parameter of distribution representing \
+        :math:`\\log\\{p/(1-p)\\}`. Either `p` or `logit` (not both) must \
+        have a value.
+
+    """
+
+    def __init__(self, p=None, logit=None):
+        super(Bernoulli, self).__init__()
+        if not (p is None) ^ (logit is None):
+            raise ValueError(
+                "Either `p` or `logit` (not both) must have a value.")
+
+        with chainer.using_config('enable_backprop', True):
+            if p is None:
+                self.logit = chainer.as_variable(logit)
+                self.p = sigmoid.sigmoid(self.logit)
+            else:
+                self.p = chainer.as_variable(p)
+                self.logit = exponential.log(self.p) - \
+                    exponential.log(1. - self.p)
```

Comment:
To avoid casting, `log1p(-self.p)` can be used.  Maybe you'd like to cache `p.dtype.type(1) - p` in the both cases.